ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -564,15 +565,18 @@ protected AstNode visitAliasedRelation(final AliasedRelation node, final Void co
       }
 
       final Optional<AliasedDataSource> existing = analysis.getSourceByName(source.getName());
-      if (existing.isPresent()) {
+      if (existing.isPresent()
+          && source.getDataSourceType().equals(DataSourceType.KTABLE)
+          && existing.get().getDataSource().getDataSourceType().equals(DataSourceType.KTABLE)) {
+
         final String errorMsg = analysis.getAllDataSources().size() > 1
-            ? "N-way joins do not support multiple occurrences of the same source. "
+            ? "N-way table-table joins do not support multiple occurrences of the same source. "
```

Comment:
Table-table self-joins are not supported because there is extra work in Streams that needs to be done to merge table sources. I.e. they are not permitted yet in Streams either